### PR TITLE
create a rope class and deepseek rope inherit from it

### PR DESCRIPTION
**Pull request by Chenyaaang**
# Description

Implement original rotary positional embedding and specific rope for deepseek. 

# Tests

Unit tests verify the cached sin/cos matrix are as expected shape, the output from rope are as expected shape and different from the original input. 

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have made or will make corresponding changes to any relevant documentation.
- [X] I have reviewed the redesigned uLLM playbook.

### CODE REVIEW

File: tpu_commons/models/jax/common/rope.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
+import math
+
+import jax
+from jax import numpy as jnp
+
+
+class RotaryEmbedding:
```
Comment (gpolovets1) [inline tpu_commons/models/jax/common/rope.py:7]: Thanks for porting this into our code! Curious if you could do a quick sanity check to see if you get the same quality/perf when pointing the models in our dashboard to this file instead of tpu_commons/tpu_commons/models/jax/layers/rope.py.
You can push the changes and run build kite as explained in the playbook (and then revert them back before we merge your PR): https://docs.google.com/document/d/1DGQBVvr2bh4G8tBUO1YH8pO7Dd_myw5rfEMfVDymEk8/edit?resourcekey=0-V7MGHu3aQjJH6YrI3-y8Hg&tab=t.0#heading=h.j4pastholmg5

File: tpu_commons/models/jax/common/rope.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        t = jnp.arange(self.original_max_position_embeddings *
+                       self.scaling_factor,
+                       dtype=jnp.float32)
+        freqs = jnp.einsum("...T,k->...Tk", t, inv_freq)
+        sin, cos = jnp.sin(freqs) * self.mscale, jnp.cos(freqs) * self.mscale
+        cache = jnp.concatenate((cos, sin), axis=-1)
+        return cache
+
+
+def _yarn_get_mscale(scale, mscale):
```
Comment (gpolovets1) [inline tpu_commons/models/jax/common/rope.py:122]: Would it be possible to use more informative naming and/or documentation/references? It's a bit hard to follow what/why the different expressions are being performed.

File: tpu_commons/models/jax/common/rope.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        return cache
+
+    def apply_rope(self, positions: jax.Array, x: jax.Array):
+        # positions should be 1d according to vllm.
+        assert x.ndim == 3
+        cos_sin = self.sin_cos_cache[positions]
+        cos, sin = jnp.split(cos_sin, 2, axis=-1)
+        assert sin.ndim == 2 and cos.ndim == 2
+        cos, sin = cos[:, None, :], sin[:, None, :]
+        x1, x2 = x[..., ::2], x[..., 1::2]
```
Comment (gpolovets1) [inline tpu_commons/models/jax/common/rope.py]: Should this actually be x[..., :x.shape[-1]//2] and x[..., x.shape[-1]//2:]? (https://github.com/AI-Hypercomputer/maxtext/blob/main/MaxText/layers/embeddings.py#L307)

File: tests/models/jax/common/test_rope.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        self.assertTrue(
+            rope.sin_cos_cache.shape == (original_max_position_embeddings,
+                                         head_dim))
+        num_tokens = 100
+        num_heads = 8
+        positions = jnp.arange(num_tokens)
+        x = jnp.ones((num_tokens, num_heads, head_dim))
+        x_rope = rope.apply_rope(positions, x)
+        self.assertTrue(x_rope.shape == x.shape)
+        self.assertFalse(jnp.array_equal(x_rope, x))
```
Comment (gpolovets1) [inline tests/models/jax/common/test_rope.py]: I think these tests may not necessarily validate the correctness of the RoPE computations. Maybe you can either hand-calculate a simple example, try find an app/implementation online that can calculate it for you, or ask Gemini.
You can also test individual components/calculations of the rope implementations to make calculations easier (it may require breaking your calculations a bit more into functions).

File: tpu_commons/models/jax/common/rope.py (hunk 3 above)
Comment (bzgoogle) [reply tpu_commons/models/jax/common/rope.py]: the interleaved partition is used in Llama4, and we will need it soon, but I would think use a traditional partition (:h/2) (h/2:h) as the default RopE?

Comment (bzgoogle) [inline tpu_commons/models/jax/common/rope.py]: nit: could we rename the x1,x2 as the first_part, first_half etc, as used in maxtext or the legacy rope?

File: tpu_commons/models/jax/common/rope.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+                           t,
+                           inv_freq,
+                           precision=jax.lax.Precision.HIGHEST)
+        sin, cos = jnp.sin(freqs), jnp.cos(freqs)
+        cache = jnp.concatenate((cos, sin), axis=-1)
+        return cache
+
+    def apply_rope(self, positions: jax.Array, x: jax.Array):
+        # positions should be 1d according to vllm.
+        assert x.ndim == 3
```
Comment (bzgoogle) [inline tpu_commons/models/jax/common/rope.py]: for the apply_rope, as we check the dim constrains for x and sin, cos, could we add shape suffix to the major variables, like x, cos, sins, x1, x2? like x -> x_TNH?

File: tpu_commons/models/jax/common/rope.py (hunk 3 above)
Comment (Chenyaaang) [reply tpu_commons/models/jax/common/rope.py]: Actually I think both are fine, as long as q and k are applied on the same rope.

File: tpu_commons/models/jax/common/rope.py (hunk 2 above)
Comment (Chenyaaang) [reply tpu_commons/models/jax/common/rope.py:122]: Actually I referred to a few implementations and they all referred from Peng et al. github.com/jquesnelle/yarn, which also doesn't have name/documentations on all functions, but I just added as much as I can. 

my referral1: https://github.com/gpolovets1/jax-llm-examples/blob/v6e_32_prof/deepseek_r1_jax/deepseek_r1_jax/model.py#L684
my referral2: https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/rotary_embedding.py#L799

File: tpu_commons/models/jax/common/rope.py (hunk 1 above)
Comment (Chenyaaang) [reply tpu_commons/models/jax/common/rope.py:7]: what are "the models in our dashboard"?

File: tpu_commons/models/jax/common/rope.py (hunk 3 above)
Comment (Chenyaaang) [reply tpu_commons/models/jax/common/rope.py]: If we select as interleaved, are they still called first/second?

Comment (gpolovets1) [reply tpu_commons/models/jax/common/rope.py]: I think your code divides x1 and x2 into even and odd entries respectively while we actually need x1 and x2 to be the first and second halves.

File: tpu_commons/models/jax/common/rope.py (hunk 2 above)
Comment (gpolovets1) [reply tpu_commons/models/jax/common/rope.py:122]: Sure, I just mean if you are pretty familiar with the calculations now, please consider adding more explanation. But if some of it you did not spend too much time understanding then we can leave it as is.

File: tpu_commons/models/jax/common/rope.py (hunk 1 above)
Comment (gpolovets1) [reply tpu_commons/models/jax/common/rope.py:7]: Chatted offline - Beinuo added our Llama3 implementation to build kite so you should probably point the attention code to this rope implementation and confirm that everything still works as expected: https://github.com/vllm-project/tpu_commons/pull/179
I believe the build kite output will also have accuracy and perf numbers so we should make sure that we don't introduce regressions there.

File: tpu_commons/models/jax/common/rope.py (hunk 3 above)
Comment (Chenyaaang) [reply tpu_commons/models/jax/common/rope.py]: ok, fixed.

File: tpu_commons/models/jax/common/rope.py (hunk 5 above)
Comment (Chenyaaang) [reply tpu_commons/models/jax/common/rope.py]: Added in comments.

File: tests/models/jax/common/test_rope.py (hunk 4 above)
Comment (Chenyaaang) [reply tests/models/jax/common/test_rope.py]: Thanks, I hand-calculate simple examples.

File: tpu_commons/models/jax/common/rope.py (hunk 1 above)
Comment (Chenyaaang) [reply tpu_commons/models/jax/common/rope.py:7]: This is currently blocked as I still don't have access to initialize a buildkite. Can any of you help me trigger it, or I can wait until I have access.

Comment (gpolovets1) [reply tpu_commons/models/jax/common/rope.py:7]: Sure, I just triggered it.

File: tests/models/jax/common/test_rope.py (hunk 4 above)
Comment (gpolovets1) [reply tests/models/jax/common/test_rope.py]: Thanks a lot this is very helpful!
